src/resourceMeta.ts

```diff
@@ -17,8 +17,9 @@ export interface Meta {
     versionId: string;
     lastUpdated: string;
     source?: string;
-    profile?: Coding;
-    tag?: Coding;
+    profile?: any | any[];
+    security?: Coding | Coding[];
+    tag?: Coding | Coding[];
 }
 
 export function generateMeta(vid: string, lastUpdatedDate: Date = new Date()): Meta {

```

Comment:
Do we need to update this method to allow customers to pass in custom meta data fields?

Reply:
I was on the fence about this - What I am implementing on the persistence side is just add this return value to the existing metadata, like:
```ts
        item.meta = { ...(item.meta ?? {}), ...generateMeta(vid.toString()) };
```

Reply:
I will also add that is give us flexibility to allow the customer to add their own metadata elements (i.e. tenants)

Reply:
In my opinion, it seems inconsistent to have a method called `generateMeta` and then to add additional metadata to it. Also it introduce the possibility that the client might build the metadata incorrectly, for example if they accidentally mess up the order of the meta item like this.
`        item.meta = { , ...generateMeta(vid.toString()), ...(item.meta ?? {}) };`

It seems safer to add an optional field to `generateMeta` like below
`generateMeta(vid: string, lastUpdatedDate: Date = new Date(), optionalFields?: any)`

Then we can unit test that this method handle all cases correctly, as in the client can't override the `vid` or `lastUpdatedTime`

Reply:
I like the approach of generating `{versionId, lastUpdated}` and merging it to the `item.meta`

I'm not sure why this method is part of the interface. It's a trivial method and only the persistence package uses it. It'd be easier if it were a private method within the persistence package. Then you could more easily rename/refactor it if `generateMeta` doesn't sound/works right anymore. 

I find the proposed `optionalFields` param to be a bit cumbersome to use. The thing is that `meta` is just another field of the resource JSON and in general we just pass JSON resources through to/from DDB without any modifications.
It'd feel redundant to manually extract `{profile, security, etc}` from `item.meta` and pass it to `generateMeta` as `optionalFields` just to have it merged back into the exact same object again.

Reply:
Actually, I think that this can be fixed without any changes to the interface.

The persistence package sometimes merges meta: https://github.com/awslabs/fhir-works-on-aws-persistence-ddb/blob/2ecc1cd894c9b10b984598f654654a92a1ae5c50/src/dataServices/dynamoDbUtil.ts#L37-L43

and sometimes overwrites meta:
https://github.com/awslabs/fhir-works-on-aws-persistence-ddb/blob/14a254e7c290b459660506c637de4601a0c36aa8/src/dataServices/dynamoDbDataService.ts#L115

we just need to change it to always merge

Reply:
Hmmm, I agree it does seem redundant to parse the `optionalFields` to `generateMeta` and then return that field back. However, doing this will help ensure the `metadata` values we're storing in DDB is internally consistent. The field `meta.profile` and `meta.tag` is supposed to have the type `Coding`. When the requester pass that value to us, we need to check those fields are valid before storing it in DDB. This validation logic needs to go somewhere. Therefore let's put it in `generateMeta` so we can reuse it. Another benefit to putting all the `meta` creation logic in `generateMeta` is that it becomes much easier to refactor any code associated with `metadata` in the future. You'll know that any `metadata` that is going to be stored in DDB will go through `getMeta` first and will be valid.

Reply:
> This validation logic needs to go somewhere. 
* This should be done in the Router/Validator; Where this method is currently being used (persistence) validation has already happened and doesn't and shouldn't happen again.

> Therefore let's put it in generateMeta so we can reuse it. 
* The goal around generateMeta is to be used across many different packages, but to Nestor's point it is a simple method and not sure if it adds much; (much like our own "clone")

Reply:
Ah, I see. Good point about meta already being validated in the Router/Validator. Given that, it makes sense to not change the interface.